openshift ex diagnostics

sosiouxme opened the pull request:
I would like some feedback on this, just not sure who to tap. @danmcp @jwhonce PTAL or redirect?

This is still mainly framework implementation:
https://trello.com/c/LdUogKuN/379-introduce-openshift-v3-diagnostics

But will actually detect a few things. (See epic for where it's going - https://trello.com/c/On7sLHVw/342-epic-openshift-v3-diagnostics)

I'd like to finally get this in front of people so they can use and improve it.

sosiouxme commented:
[test]

danmcp commented on pkg/diagnostics/README.md:
```diff
@@ -0,0 +1,51 @@
+under heavy development in parallel with the OpenShift Origin project.
+
+The goals of the diagnostics tool are summarized in this
+[Trello card](https://trello.com/c/LdUogKuN/379-openshift-v3-diagnostics)
```
Should use the short form of the link.

danmcp commented on pkg/diagnostics/README.md:
```diff
@@ -0,0 +1,51 @@
+
+The goals of the diagnostics tool are summarized in this
+[Trello card](https://trello.com/c/LdUogKuN/379-openshift-v3-diagnostics)
+(account required). Eventually the functionality may be included more
```
What account is required?

sosiouxme replied:
Oh geez, I entirely forgot to update this README. Thanks, I'll do that.

sosiouxme commented:
@jwhonce PTAL or redirect? 

Dan suggested that having the large explanation text in the middle of the code makes it a bit unreadable and these could be separated out. I did kind of have in mind doing that at some point, but on the other hand I like having things in context. Dan declined to comment on code quality, deferring to someone more Go-focused.

Comments on that, how idiomatic the code is(n't), or really anything - all welcome.

sosiouxme commented:
[test] again after v0.5 rebase

nak3 commented on pkg/diagnostics/systemd/systemd.go:
```diff
@@ -0,0 +1,547 @@
+var tlsClientErrorSeen map[string]bool
+
+// Specify what units we can check and what to look for and say about it
+var unitLogSpecs = []*unitSpec{
```
gofmt test doesn't pass. https://travis-ci.org/openshift/origin/jobs/61195404
You should run `./hack/verify-gofmt.sh` and fix them.

sosiouxme replied:
Thanks, fixed it.

nak3 replied:
Sorry, I guess I commented the line which would not be fixed. Please delete the comment and close the code expansion.

jwhonce commented:
@pmorie or/and @mfojtik Can  one or both of you please review for proper "go-ness" in the PR.  Thanks.

sosiouxme commented:
FYI @deads2k @liggitt since you'll be likely to bump up against diagnostics with any config changes...

deads2k commented on hack/common.sh:
```diff
@@ -56,6 +56,7 @@ readonly OPENSHIFT_BINARY_SYMLINKS=(
   openshift-sti-build
   openshift-docker-build
   openshift-gitserver
+  openshift-diagnostics
```
If you're going to have this, I say shorter name.  At that length, it may as well just be `openshift diagnostics`.

sosiouxme replied:
It tab-completes though and so is discoverable :)

I agree the name is too long but I didn't really like any of the alternatives and didn't hear a consensus. I figure that might develop after people actually use it.

deads2k commented on hack/test-end-to-end.sh:
```diff
@@ -358,6 +358,13 @@ echo "[INFO] Validating exec"
+# show only errors, fail if any errors
+wait_for_command "openshift-diagnostics --config=${OPENSHIFTCONFIG} --master-config=${MASTER_CONFIG_DIR}/master-config.yaml --node-config=${NODE_CONFIG_DIR}/node-config.yaml --diaglevel=0" 60
+wait_for_command "openshift ex diagnostics client --config=${OPENSHIFTCONFIG} --diaglevel=0" 15
+sleep 300
```
Why is this here?

sosiouxme replied:
Errr... because I'm a doofus

deads2k commented on hack/test-end-to-end.sh:
```diff
@@ -358,6 +358,13 @@ echo "[INFO] Validating exec"
+# Diagnostics
+echo "[INFO] Running openshift-diagnostics"
+# show only errors, fail if any errors
+wait_for_command "openshift-diagnostics --config=${OPENSHIFTCONFIG} --master-config=${MASTER_CONFIG_DIR}/master-config.yaml --node-config=${NODE_CONFIG_DIR}/node-config.yaml --diaglevel=0" 60
```
This takes 60 seconds to run?

deads2k replied:
`--config=${OPENSHIFTCONFIG}` should not be necessary.  That ought to happen by default.

sosiouxme replied:
Shouldn't take 60 seconds. Just didn't want the test to randomly fail because you happen to have a lot of docker journal messages to analyze or something.

`openshift-diagnostics client` follows standard config loading. Plain `openshift-diagnostics` does not - it expects that config files are either specified or in a standard location. We can talk about this as I can see how it's confusing but divining user intent is tricky here...

sosiouxme replied:
I'll revisit client config loading. It may already work or there is likely something reasonable to do here.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,214 @@
+intended to be run from the same context as an OpenShift client or running
+master / node in order to troubleshoot from the perspective of each.
+
+    $ %s
```
`%[1]s` ?

sosiouxme replied:
yes

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,214 @@
+can be used with subcommands.
+`
+
+func NewCommand(name string) *cobra.Command {
```
Take `name`, `fullName`, and `Out`

mfojtik commented on pkg/diagnostics/client/client.go:
```diff
@@ -0,0 +1,145 @@
+			if _, kclient, err := env.ClusterAdminFactory.Clients(); err == nil {
+				nodes, err = kclient.Nodes().List(labels.LabelSelector{})
+			}
+			if err != nil {
```
you can get rid of the global 'err' variable if you use 'else' here

sosiouxme replied:
As written it also addresses the case where err is set from listing nodes (inside the if)

deads2k commented on pkg/diagnostics/types/flags.go:
```diff
@@ -0,0 +1,53 @@
+	OscPath        string
+	LogLevel       int
+	Format         string          // format of output - text/json/yaml
+	OpenshiftFlags *flag.FlagSet   // flags that were set on the command run
```
Try to keep these logically distinct.  We've been splitting commands into stages:
 1. Gather the data required for the command into an intermediate (`Options`) object.  This is done via binding or in `.Complete`
 1. Validate the data given for a command. `.Validate`
 1. Run the guts of the command based on the `Options`. 

This approach makes it easier to write unit tests and makes it possible to delegate from one command to another with minimal pain.  As examples, you can see our usage in `kubectl config` (testing), `osc login` and `osc project` (delegatees), and `openshift start` (delegator).  We also re-use these commands in integration testing, so there's more examples there.

deads2k commented on pkg/diagnostics/types/flags.go:
```diff
@@ -0,0 +1,53 @@
+	}
+}
+
+type List []string
```
`StringList` already exists in kube's `util` package.

deads2k commented on pkg/diagnostics/types/flags.go:
```diff
@@ -0,0 +1,53 @@
+	// the master has its own container for flags/options
+	MasterOptions *start.MasterOptions
+	// the node has its own container for flags/options
+	NodeOptions *start.NodeOptions
```
This feels gorpy.  Looking at the package, it seems like this set object gets passed to commands even if those commands aren't supposed to check the Node.  What does it look like to have the information required for each command split apart?  Using partial objects makes it harder to reason about what I actually need to set.

sosiouxme replied:
The interplay is a little complicated. The same command can be checking any/all of master,node,client. Some options are shared between those (e.g. syslogd unit status between node/master; list of named diagnostics you want to run), some aren't.

I think it will probably be possible/useful to more clearly delineate these as separate objects as you say. At a minimum, one for handling the top-level "what did the user say and what is discoverable" and one each for handing off to actually run master,node,client diagnostics.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+// Used in the command definition for common flags that we want to be visible in usage
+func addFlags(cmd *cobra.Command, flags *types.Flags) {
```
Move the bind method closer to the type being bound and use an object to allow overriding of the flags.  That leaves it easy to bind the defaults for consistency, but possible to re-use the flags with slightly different names.  See https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/clientcmd/overrides.go as an example.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,214 @@
+// Every command invocation needs to do the same things at the beginning...
+func runInit(cmd *cobra.Command, diagFlags *types.Flags) {
+	cmd.SetOutput(os.Stdout)                         // for output re usage / help
+	diagFlags.OpenshiftFlags = cmd.PersistentFlags() // capture flags from *this* command in env
```
Didn't you already set this?  I guess you might have added more if you're calling from other commands, but in that case why not use a merged list?  This will be clearer to read after splitting up `types.Flags` and divorcing it from cobra and flags.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,214 @@
+func runInit(cmd *cobra.Command, diagFlags *types.Flags) {
+	cmd.SetOutput(os.Stdout)                         // for output re usage / help
+	diagFlags.OpenshiftFlags = cmd.PersistentFlags() // capture flags from *this* command in env
+	log.SetLevel(diagFlags.LogLevel)
```
Side-effecty.  What is this even setting?

sosiouxme replied:
Log level. Sure it's side-effecty. Logging is side-effecty. Before it has even determined anything there is stuff to output, and it has to know whether it's of sufficient log level to output, and what format to use for the log.

Probably a logger object can be used as you say.

deads2k commented on pkg/diagnostics/log/log.go:
```diff
@@ -0,0 +1,197 @@
+package log
```
Since you have settings, you should be passing through instances of Log on your options object.  In delegation cases, I'm going to want strong control over exactly what comes out.  This the second biggest failing of `glog`.

deads2k replied:
Put another way, this is a stateful output utility and stateful things should be instance wide, not execution environment wide.

sosiouxme replied:
... So, now loggers are instances.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,214 @@
+    $ %s
+`
+
+func NewClientCommand(fullName string, diagFlags *types.Flags) (*cobra.Command, *osclientcmd.Factory) {
```
accept `name` as well.  In embedding cases, name might reasonably change.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,214 @@
+
+	// Add some diagnostics flags to be shown separately from client flags
+	addFlags(cmd, diagFlags)
+	cmd.Flags().StringVarP(&diagFlags.OpenshiftPath, "openshift", "", "", "Path to 'openshift' binary")
```
Huh?  Can you update the help to make it a little more obvious why I would need or want to tell you where my openshift binary is?

sosiouxme replied:
The idea was that for a side-loaded diagnostics command (which would be likely at least for a while) it would be good to know if there was a difference in versions between the openshift binary in use and the one doing the diagnostics (so if running diagnostics from origin versus an enterprise install a few months back, we would expect different capabilities, or just warn about the difference). Also it would be useful to find in a process list.

But several factors have emerged that probably make this fairly irrelevant now.

deads2k commented:
Can you refactor into something like I've described here: https://github.com/openshift/origin/pull/2049#discussion_r29847589 ?  I'm having some trouble parsing the chain of events in the command and I think that separation will help make it easier for me to reason about what's happening.

Stop by if you want to talk about it some more.  I'll pick up the review after some of the initial bits are cleared up.

sosiouxme commented:
Going to rework this some.

sosiouxme commented:
@deads2k I think this is a good bit closer to what you were describing. Command flag definitions and what they bind into are cleanly delineated. Loggers are stateful objects passed in.

I considered breaking up discovery.Environment along similar lines, but that seemed less clearly delineated.

Anyway, your feedback would be most appreciated.

deads2k commented on .gitignore:
```diff
@@ -9,6 +9,7 @@
 *.swp
 .vimrc
 .kubeconfig
+.openshiftconfig
```
This file doesn't normally exist.  Where did you get it from?

liggitt replied:
I like excluding it... we'll never want someone to check it in

liggitt replied:
I retract my comment... `./.openshiftconfig` is not a known filename we generate

sosiouxme replied:
There was a point where client looked in ./ for config. Handy for testing after compiling on my own machine instead of some slow cloud VM.

deads2k replied:
> There was a point where client looked in ./ for config. Handy for testing after compiling on my own machine instead of some slow cloud VM.

It doesn't anymore.  Delete it and `.kubeconfig`

sosiouxme replied:
Git blame says:

```
d7e40728 (deads2k           2015-02-02 10:14:58 -0500 11) .kubeconfig
```

:-P will do

deads2k commented on docs/debugging-openshift.md:
```diff
@@ -25,6 +25,41 @@ System Environment
+or specified. For any (potential) problems that it detects, it will explain
+in detail what is going on.
+
+    $ openshift ex diagnostics
```
`openshift-diagnostics`

sosiouxme replied:
It's both!

deads2k replied:
> It's both! And in this case I kind of want to emphasize that it's experimental AKA alpha

If you aren't ready for it to be exposed, don't expose it as `openshift-diagnostics`.  I don't have a preference one way or the other, but things expose as main commands are referenced by their official names and only live under `ex` for compatibility until we clean them up.

sosiouxme replied:
That's fair. Won't expose top-level command then.

deads2k commented on docs/debugging-openshift.md:
```diff
@@ -25,6 +25,41 @@ System Environment
+
+In any system like OpenShift where a lot of components are working together, 
+it can be difficult to figure out where a process is breaking down. To help
+with this discovery and diagnosis, `openshift ex diagnostics` can detect a
```
`openshift-diagnostics`

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,198 @@
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/spf13/cobra"
+	"io"
+	//"github.com/openshift/origin/pkg/cmd/cli/cmd"
```
delete cruft

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,198 @@
+You may also specify config files explicitly with flags below, in which
+case you will receive an error if they are not found or invalid.
+
+    $ %[1]s --master-config=/etc/openshift/master.yaml
```
@sdodson is that our default location?

sosiouxme replied:
aw crap it's changed now. thanks.

sdodson replied:
yeah, it's /etc/openshift/master/master-config.yaml now

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,198 @@
+use the same flags that are actually set on the command line for that
+component to configure the diagnostic.
+
+    $ %[1]s node --master=https://master.openshift.example.com:8443 --cert-dir=...
```
Flag names should be consistent.  Shouldn't this be `--server`?

sosiouxme replied:
actually according to `openshift start node -h` it no longer takes either... takes --kubeconfig

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,198 @@
+	   to add those flags to this command (the only client option here is a config
+	   file). So the factory object from client cmd is reused for this command.
+	*/
+	ccmd, factory := NewClientCommand("client", name+" client", out)
```
better name than `ccmd`.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,198 @@
+	   This command needs the client factory built in the "client" subcommand.
+	   Generating the factory adds flags to the "client" cmd, and we do not want
+	   to add those flags to this command (the only client option here is a config
+	   file). So the factory object from client cmd is reused for this command.
```
Wait, what?  There's probably a better way.  I'll think about it during the review.

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -102,23 +103,28 @@ func NewCommandStartMaster(out io.Writer) (*cobra.Command, *MasterOptions) {
 	return cmd, options
 }
 
+func MasterArgsAndFlags(flags *pflag.FlagSet) *MasterArgs {
```
You can group the `Binds` together, but only if you do all of them (see the other bindings still in `NewCmd`) and you need to keep the method separate from creating the MasterArgs themselves.

sosiouxme replied:
Would you prefer I just copy and paste the binds that make sense?

They don't all make sense elsewhere. You would not write master config or certs from diagnostics. But by sharing this method, diagnostics keep up with config drift for free.

deads2k replied:
I would make the case that someone likely to run diagnostics is also likely to run from config, so you probably only need `--master-config` and `--node-config`

sosiouxme replied:
I think this should be useful even in a self-compiled run-with-options setup. It may not seem like there's much to do with a master or node from this setup, since you have no config to check and probably no logs (unless we introduce an option to specify where to find those and scrape them in addition to journald) - but you can at least check systemd dependencies like Docker, possibly etcd or kubernetes, which could be useful.

liggitt replied:
If it was trivially straightforward to support all the same flags, I wouldn't care much. If it's gorpy and makes two places we have to maintain, I'd say just diagnose from config

sosiouxme replied:
I pulled out the shared flags in a separate method so they would be straightforward to support...

deads2k commented on pkg/cmd/server/start/start_node.go:
```diff
@@ -77,17 +78,20 @@ func NewCommandStartNode(out io.Writer) (*cobra.Command, *NodeOptions) {
-	BindListenArg(options.NodeArgs.ListenArg, flags, "")
-	BindImageFormatArgs(options.NodeArgs.ImageFormatArgs, flags, "")
-	BindKubeConnectionArgs(options.NodeArgs.KubeConnectionArgs, flags, "")
+func NodeArgsAndFlags(flags *pflag.FlagSet) *NodeArgs {
```
Same comment as in master.

sosiouxme commented:
I squashed, rebased, added a registry diagnostic.

sosiouxme commented on pkg/diagnostics/systemd/analyze_logs.go:
```diff
@@ -0,0 +1,128 @@
+func (d AnalyzeLogs) Check() (bool, []log.Message, []error, []error) {
+	infos := []log.Message{}
+	warnings := []error{}
+	errors := []error{}
```
@deads2k finally resurrecting this. I think I want to refactor the actual logging entirely out of the diagnostics themselves. This whole signature and intro are so ugly and awkward to interact with, and they don't even capture all the different log levels available or the order in which infos/errors/warnings are found.

Instead I think I want to return a *DiagnosticResult (new type). The role of the diagnostic would be to put messages, warnings, and errors into this result and set success. Then the thing running the diagnostic can decide what to actually log and when. This would help with parallelization too. That means you don't see partial logs produced during the run of an individual diagnostic, but I think I can live with that (and with parallel execution, that would have been semi-necessary anyway).

What do you think?

deads2k replied:
> Instead I think I want to return a *DiagnosticResult (new type). The role of the diagnostic would be to put messages, warnings, and errors into this result and set success.

Yes, I like that idea.  I see two potential problems, but I think they're surmountable.
 1. Inability to show partial output.  If a diagnostic test gets stuck, you can't see the errors it produced along the way.
 1. Inability to interleave.  Some output is related to other output.  For instance, if you're checking a particular kubeconfig context and find two warnings and an error, you often want to display the error near the warnings.

Perhaps a DiagnosticResult could contain something like these markers to allow sorting? https://github.com/openshift/origin/blob/master/pkg/api/graph/interfaces.go#L8  Having `originator`, `severity`, `key`, `message`, `advice` would structure the output and solve problem 2.

I'd probably add high verbosity glog statements to deal with problem 1.

sosiouxme replied:
Good ideas all, will work on it today.

sosiouxme commented:
@deads2k just wanted to give you a heads up that completing this is finally on the radar so you can review soon (or pawn off on someone else if you're sick of it). I still have some cleanup - dead code, adding back diagnostics that got left behind in the refactor, TODOs, that sort of thing, but I think the bones are where I want them; interested to see how it looks to you. In the end, I saw the wisdom of diagnostics being neatly compartmentalized with just what they need to run. Thanks.

sosiouxme commented:
Also the log analysis is now limited so it shouldn't run forever :)

mfojtik commented on pkg/cmd/experimental/diagnostics/client.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+const (
+	ConfigContexts = "ConfigContexts"
```
godoc

mfojtik commented on pkg/cmd/experimental/diagnostics/client.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+var (
+	AvailableClientDiagnostics = util.NewStringSet(ConfigContexts) // add more diagnostics as they are defined
```
godoc (otherwise you will make golint unhappy ;-)

Also is it necessary to have this variable public?

sosiouxme replied:
Since it's really only used in the package, I suppose it doesn't need to be public. If it's needed for integration tests or something later, we can make it that way.

mfojtik commented on pkg/cmd/experimental/diagnostics/client.go:
```diff
@@ -0,0 +1,45 @@
+	AvailableClientDiagnostics = util.NewStringSet(ConfigContexts) // add more diagnostics as they are defined
+)
+
+func (o DiagnosticsOptions) buildClientDiagnostics(rawConfig *clientcmdapi.Config) ([]types.Diagnostic, bool /* ok */, error) {
```
move the /* ok */ to godoc (if you want to tell what that boolean return)

mfojtik commented on pkg/cmd/experimental/diagnostics/client.go:
```diff
@@ -0,0 +1,45 @@
+func (o DiagnosticsOptions) buildClientDiagnostics(rawConfig *clientcmdapi.Config) ([]types.Diagnostic, bool /* ok */, error) {
+
+	osClient, kubeClient, clientErr := o.Factory.Clients()
+	_ = osClient   // remove once a diagnostic makes use of OpenShift client
```
if we don't use this now, can't you just comment it?

mfojtik commented on pkg/cmd/experimental/diagnostics/cluster.go:
```diff
@@ -0,0 +1,115 @@
+		return nil, true, nil // don't waste time on discovery
+	}
+
+	var clusterClient *client.Client
```
group this:

```
var (
 ...
)
```

mfojtik commented on pkg/cmd/experimental/diagnostics/cluster.go:
```diff
@@ -0,0 +1,115 @@
+	diagnostics := []types.Diagnostic{}
+	for _, diagnosticName := range requestedDiagnostics {
+		switch diagnosticName {
+		case "NodeDefinitions":
```
I would make this `const`

mfojtik commented on pkg/cmd/experimental/diagnostics/cluster.go:
```diff
@@ -0,0 +1,115 @@
+	if o.ClientClusterContext != "" { // user has specified cluster context to use
+		if context, exists := rawConfig.Contexts[o.ClientClusterContext]; exists {
+			configErr := fmt.Errorf("Specified '%s' as cluster-admin context, but it was not found in your client configuration.", o.ClientClusterContext)
+			o.Logger.Error("discClustCtx", configErr.Error())
```
provide more readable error

sosiouxme replied:
I'm open to idea's on what's more readable than "Specified '%s' as cluster-admin context, but it was not found in your client configuration."

Granted people may not know much about config and contexts, but is this the place to explain it all? Maybe link to docs?

deads2k replied:
> I'm open to idea's on what's more readable than "Specified '%s' as cluster-admin context, but it was not found in your client configuration."
> 
> Granted people may not know much about config and contexts, but is this the place to explain it all? Maybe link to docs?

He's talking about "discClustCtx".  

@mfojtik I think the intent here is for that funny string to be an ID associated with particular errors to allow eventual `grep` or key based advice.

sosiouxme replied:
If so - yeah, that's the key, not an error message. The idea is that automation could filter for particular error IDs and not have to read the text to decide whether it's important.

mfojtik commented on pkg/cmd/experimental/diagnostics/cluster.go:
```diff
@@ -0,0 +1,115 @@
+	currentContext, exists := rawConfig.Contexts[rawConfig.CurrentContext]
+	if !exists { // config specified cluster admin context that doesn't exist; complain and quit
+		configErr := fmt.Errorf("Current context '%s' not found in client configuration; will not attempt cluster diagnostics.", rawConfig.CurrentContext)
+		o.Logger.Errorf("discClustCtx", configErr.Error())
```
same here

mfojtik commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,247 @@
+	"github.com/openshift/origin/pkg/diagnostics/types"
+)
+
+type DiagnosticsOptions struct {
```
godoc

mfojtik commented on pkg/cmd/experimental/diagnostics/host.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+const (
+	StandardMasterConfigPath string = "/etc/openshift/master/master-config.yaml"
```
godoc

mfojtik commented on pkg/cmd/experimental/diagnostics/host.go:
```diff
@@ -0,0 +1,76 @@
+	systemdUnits := systemddiagnostics.GetSystemdUnits(o.Logger)
+	for _, diagnosticName := range requestedDiagnostics {
+		switch diagnosticName {
+		case "AnalyzeLogs":
```
I think this should be const

sosiouxme replied:
Yes, I'm going through and changing all the diagnostic name strings to constants.

mfojtik commented on pkg/cmd/experimental/diagnostics/options/diagnostics.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+// default overrideable flag specifications to be bound to options.
+func RecommendedLoggerOptionFlags() LoggerOptionFlags {
```
godoc (and all the rest func)

sosiouxme commented:
I'm gathering that I need to write godocs :) Thanks for the feedback.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,247 @@
+		`, log.Hash{"requested": o.RequestedDiagnostics, "common": common.List(), "available": AvailableDiagnostics.List()})
+	}
+
+	func() { // don't trust discovery/build of diagnostics; wrap panic nicely in case of developer error
```
This isn't necessary.  All of our client code just lets it fail catastrophically.

sosiouxme replied:
But that's client code. Allowing your diagnostics to crash for an end user destroys credibility.

deads2k replied:
> But that's client code. Allowing your diagnostics to crash for an end user with a bare stack trace and no context destroys credibility. Don't forget this will be the first line of support.

I'm unsold.  How about we get the non-contentious parts in first and add this back after the main body of work is merged?

deads2k commented on pkg/diagnostics/types/diagnostic.go:
```diff
@@ -0,0 +1,175 @@
+	return &DiagnosticResult{origin: origin}
+}
+
+func (r *DiagnosticResult) Complete() *DiagnosticResult {
```
do this in your constructor.

sosiouxme replied:
OK. This is a relic from when the constructor was optional.

deads2k commented on pkg/diagnostics/types/diagnostic.go:
```diff
@@ -0,0 +1,175 @@
+	return r.errors
+}
+
+func (r *DiagnosticResult) Append(r2 *DiagnosticResult) {
```
This seems fundamentally incompatible with the single `origin` above.

sosiouxme replied:
You're right. And the one place actually using this should probably do it a little differently.

deads2k commented on pkg/diagnostics/types/diagnostic.go:
```diff
@@ -0,0 +1,175 @@
+
+type DiagnosticResult struct {
+	failure  bool
+	origin   string // name of diagnostic; automatically inserted into log Entries
```
`name` so it matches the `Diagnostic` interface above.

sosiouxme replied:
It doesn't necessarily; I should fix the comment.

deads2k commented on pkg/diagnostics/types/diagnostic.go:
```diff
@@ -0,0 +1,175 @@
+
+// Public ingress functions
+// Errors are recorded as errors and also logged
+func (r *DiagnosticResult) Error(id string, err error, text string) {
```
Building a return value from a utility function is distinct from logging that information.  Try keeping this wrapper thin, `AddError(DiagnosticError)`.  I could even see exposing those fields directly on the struct.

deads2k replied:
> Building a return value from a utility function is distinct from logging that information. Try keeping this wrapper thin, AddError(DiagnosticError). I could even see exposing those fields directly on the struct.

Ok, I see.  This isn't reporting the errors.  Still keep the wrapper thin.  `AddError(DiagnosticError)`.  I think it make life easier.

sosiouxme replied:
Updating comment to:

    // Errors are recorded as errors plus log entries made in the result

This isn't just reporting an error. It's explaining and adding context, which is the whole purpose of diagnostics. And I want that to be as easy as possible.

deads2k commented on pkg/diagnostics/types/diagnostic.go:
```diff
@@ -0,0 +1,175 @@
+	"github.com/openshift/origin/pkg/diagnostics/log"
+)
+
+type Diagnostic interface {
```
I like the interface.  Add godoc.

deads2k commented on pkg/diagnostics/types/diagnostic.go:
```diff
@@ -0,0 +1,175 @@
+	Check() *DiagnosticResult
+}
+
+type DiagnosticResult struct {
```
External interaction is through methods.  I don't object to that, but if that's so, then you should the types should reflect that.  This is rightly a public interface with a private impl.

sosiouxme replied:
Alright, that's simple enough.

deads2k commented on pkg/cmd/experimental/diagnostics/host.go:
```diff
@@ -0,0 +1,76 @@
+	AvailableHostDiagnostics = util.NewStringSet("AnalyzeLogs", "UnitStatus", "MasterConfigCheck", "NodeConfigCheck")
+)
+
+func (o DiagnosticsOptions) buildHostDiagnostics() ([]types.Diagnostic, bool /* ok */, error) {
```
return of `[]types.Diagnostic, error` is probably sufficient

sosiouxme replied:
I was never really happy about this, but I do think there is a useful distinction between indicating "there was an error we should track" vs "we should abort running any diagnostics".

deads2k replied:
> I was never really happy about this, but I do think there is a useful distinction between indicating "there was an error we should track" vs "we should abort running any diagnostics".

The argument would be stronger if I saw a way to return `false, nil`

sosiouxme replied:
You wouldn't return `false, nil` as there would be no reason to abort if
there is no error. But it can (and does) return both `false, err` (there
was an error, abort) or `true, err` (there was an error, but continue
anyway).

On Tue, Aug 18, 2015 at 9:03 AM, David Eads <notifications@github.com>
wrote:

> I was never really happy about this, but I do think there is a useful
> distinction between indicating "there was an error we should track" vs "we
> should abort running any diagnostics".
>
> The argument would be stronger if I saw a way to return false, nil
>

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,247 @@
+	return failed, err, numWarnings, numErrors
+}
+
+func (o DiagnosticsOptions) Run(diagnostics map[string][]types.Diagnostic) (bool, error, int, int) {
```
Given three well known categories, let's just have three `[]types.Diagnostic` and callouts to something like `func (o DiagnosticsOptions) RunCategory(name string, diagnostics []types.Diagnostic)`

sosiouxme replied:
I can easily imagine at least two more categories springing up (in-server diagnostics, and diagnostics container). Does that change anything? Personally, whenever there are more than two of anything, I prefer to make it generic...

deads2k replied:
> I can easily imagine at least two more categories springing up (in-server diagnostics, and diagnostics container). Does that change anything? Personally, whenever there are more than two of anything, I prefer to make it generic...

I'm ok with a container object with specific fields.  Thing is, this isn't truly generic and the map only pretends to make it so.  There's a finite, small, and well known number of buckets that have distinct semantic utility.

sosiouxme replied:
On Tue, Aug 18, 2015 at 9:05 AM, David Eads <notifications@github.com>
wrote:
>
> > +func (o DiagnosticsOptions) Run(diagnostics map[string][]types.Diagnostic) (bool, error, int, int) {
>
> I can easily imagine at least two more categories springing up (in-server
> diagnostics, and diagnostics container). Does that change anything?
> Personally, whenever there are more than two of anything, I prefer to make
> it generic...
>
> I'm ok with a container object with specific fields. Thing is, this isn't
> truly generic and the map only pretends to make it so. There's a finite,
> small, and well known number of buckets that have distinct semantic
> utility.
>
I changed it to a container object.

sosiouxme replied:
On Wed, Aug 19, 2015 at 8:28 AM, Luke Meyer <lmeyer@redhat.com> wrote:
>
> I changed it to a container object.
>

And then just removed the whole concept of diagnostic areas... it wasn't
really adding much. Easy to add later if needed.

deads2k commented on pkg/cmd/experimental/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,247 @@
+	errorCount := 0
+	for area, areaDiagnostics := range diagnostics {
+		for _, diagnostic := range areaDiagnostics {
+			func() { // wrap diagnostic panic nicely in case of developer error
```
just let it die horribly.

sosiouxme replied:
And deprive the user of any further diagnostics that might actually run successfully?

deads2k commented:
@danmcp @smarterclayton Is a machine readable output from diagnostics a design goal?  It leads to more difficult to read implementations due to golang templates and maps to hold pertinent data as opposed to simple printf style strings.  I'm inclined to say that machines aren't likely to be good at presenting the information anyway, so let's punt on it.

smarterclayton commented:
I don't see it as a requirement for now.

On Aug 17, 2015, at 4:32 PM, David Eads <notifications@github.com> wrote:

@danmcp <https://github.com/danmcp> @smarterclayton
<https://github.com/smarterclayton> Is a machine readable output from
diagnostics a design goal? It leads to more difficult to read
implementations due to golang templates and maps to hold pertinent data as
opposed to simple printf style strings. I'm inclined to say that machines
aren't likely to be good at presenting the information anyway, so let's
punt on it.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/2049#issuecomment-131953624>.

sosiouxme commented:
@deads2k I don't understand what's to punt on. It's already in there. It would be extra work to remove. And I disagree that it makes the implementation substantially more difficult to read. Is this:

    r.Warnf("text string with %[1]s and %[2]s %[1]s", stuffExpression, moreExpression})

really so much better than this?

    r.Warnt("clNodeNotSched", nil, "text string with {{.stuff}} and {{.more}} {{.stuff}}", log.Hash{"stuff": stuffExpression, "more": moreExpression})

Not saying it's always preferable, but you have the choice of either.

I think sysadmins always want machine-readable output. Think of automated notification systems.

And this also gets us halfway to I18N/L10N of messages.

Why go backward?

deads2k commented on pkg/diagnostics/log/log.go:
```diff
@@ -0,0 +1,280 @@
+	Finish()
+}
+
+func NewLogger(setLevel int, setFormat string, out io.Writer) (*Logger, error) {
```
Based on https://github.com/openshift/origin/pull/2049#issuecomment-131992506, we can simplify this.  We won't need json and yaml serialization to start.  That will let you eliminate the golang template calls.  Most people find them very hard to read when compared to printf.

I think that's fine because of the `DiagnosticResult` type that could later be used to include pertinent metadata for structured results if we ever develop that far.

sosiouxme replied:
I would want to use templates even if we didn't do json and yaml. I don't understand the objection to them. They are the only sane way to write some things. Taking them out doesn't even simplify anything AFAICS.

deads2k commented:
> I think sysadmins always want machine-readable output. Think of automated notification systems.
> 
> And this also gets us halfway to I18N/L10N of messages.
> 
> Why go backward?

I've gone through that bit of code a couple time and I'm still not happy with the way it has ended up being factored out.  I suspect that what I'll end up wanting is a nearly complete implementation of JSR47 or some equivalent.  Basically, you're trying to implement the intent of configuration, logging levels, functional area trees, filters, handlers, internationalization, and formatters without abstracting each piece and building the proper linkages.  In order to build a composable system of logging, you really need those things.

I'm certainly willing to accept the one-off implementation of levels (without functional area trees) and a single handler/formatter tuple without building that infrastructure, but I think you've bitten off more than can reasonably be done in this pull.

I really want diagnostics and you really want diagnostics, but the logging factorization will hold up this pull if we try to handle it correctly.

liggitt commented on pkg/cmd/openshift/openshift.go:
```diff
@@ -52,6 +53,8 @@ func CommandFor(basename string) *cobra.Command {
 		cmd = irouter.NewCommandTemplateRouter(basename)
 	case "openshift-deploy":
 		cmd = deployer.NewCommandDeployer(basename)
+	case "openshift-diagnostics":
```
let's keep it as an experimental subcommand for now... top-level commands require install changes to actually be available. was there agreement that we wanted a top-level command for this one?

sosiouxme replied:
No, I meant to take this out already.

sosiouxme commented:
@liggitt are you waiting on anything to review? I wanted to get in any other changes needed before I rip out all the complicated logging stuff. Unless you are wanting me to do that before you review.

liggitt commented:
@sosiouxme go ahead, you'll probably finish before I get back to it

liggitt commented on .gitignore:
```diff
@@ -8,7 +8,6 @@
 /examples/sample-app/logs/openshift.log
 *.swp
 .vimrc
-.kubeconfig
```
we don't want people committing kubeconfig files, do we?

sosiouxme replied:
No, but there was a comment LOOONG ago when I was trying to *add* a gitignore for openshift config that since we don't look in pwd for config any more, there was no reason for this cruft. Doesn't matter much to me... you want to keep this?

liggitt replied:
I suppose that's true… just didn't see the relevance to the pull, and didn't consider it too crufty

liggitt commented on pkg/diagnostics/README.md:
```diff
@@ -0,0 +1,113 @@
+OpenShift server or client, the diagnostics can run in the exact same
+environment.
+
+Diagnostics looks for config files in standard locations. If not found,
```
enumerate what those are here?

sosiouxme replied:
now enumerated below

liggitt commented on pkg/diagnostics/README.md:
```diff
@@ -0,0 +1,113 @@
+Having configuration files in standard locations means you will generally
+not even need to specify where to find them. Running:
+
+    openshift ex diagnostics
```
it's a little odd for diagnostics to default to paths that `openshift start` doesn't even default to

sosiouxme replied:
You mean the standard master/node config locations?

Granted those are only "standard" in the sense that ansible puts them there. But I can't think of a better way for the user not to have to specify these most of the time...

liggitt replied:
just saying, there is nothing else in the codebase that looks there, not even the command to run the server

sosiouxme replied:
there is nothing in the codebase to look for a master/node config at all... you have to specify it with a flag.   
But FWIW the packaging (though not the code) in this repo does refer to these standard locations.
E.g. https://github.com/openshift/origin/blob/master/rel-eng/origin-master.sysconfig#L2

liggitt commented on pkg/diagnostics/README.md:
```diff
@@ -0,0 +1,113 @@
+messages should:
+
+* Have an ID that is unique and unchanging, such that automated alerts
+  could filter on specific IDs rather than rely on message text or level.
```
any ideas for ensuring a unique message id?

liggitt replied:
since we're claiming that ids won't change, I'd like to get the message ids looking a little better... they're tantalizingly close to being understandable at the moment:
```
discCCnotFlag
clRegISMismatch
clRtPodConn
discMCfound
sdLogOMreBadCert
```

seems like a good pattern might be "parentTest-messageID"? Is there a reason not to make the message ids use full words?
```
discCCnotFlag -> configLoading-missingFlagDiscovered
clRegISMismatch -> clusterRegistry-imageStreamMismatch
clRtPodConn -> clusterRegistry-podAPIConnection
discMCfound -> masterConfigCheck-discoveredConfig
sdLogOMreBadCert -> systemdLog-masterTLSHandshakeBadCertificate
```

sosiouxme replied:
Because they'd be long, and someone has to type them in the source, and they show up in the output? And I can't hit '*' in vim to check for uniqueness? *shrug* They're only semi-readable to make it easier to come up with unique ones. They could all be E12351. I didn't have any bright ideas for ensuring uniqueness, until we I18Nize everything and they're message lookups. Without machine-directed output they're less interesting. If you feel they should be more readable I can do that.

liggitt replied:
I'd be fine with pure IDs (E012345) as long as we ensured uniqueness, or
more readable ones. Halfway between is uncomfortable.



On Aug 30, 2015, at 8:24 PM, Luke Meyer <notifications@github.com> wrote:

In pkg/diagnostics/README.md
<https://github.com/openshift/origin/pull/2049#discussion_r38278279>:

> +paths or flooding them with non-issues that obscure real problems.
> +
> +* Errors should be reserved for things that are almost certainly broken
> +  or causing problems, for example a broken URL.
> +* Warnings indicate issues that may be a problem but could be valid for
> +  some configurations / situations, for example a node being disabled.
> +
> +Enabling automation
> +===================
> +
> +Diagnostic messages are designed to be logged either for human consumption
> +("text" format) or for scripting/automation ("yaml" or "json" formats). So
> +messages should:
> +
> +* Have an ID that is unique and unchanging, such that automated alerts
> +  could filter on specific IDs rather than rely on message text or level.

Because they'd be long, and someone has to type them in the source, and
they show up in the output? And I can't hit '*' in vim to check for
uniqueness? *shrug* They're only semi-readable to make it easier to come up
with unique ones. They could all be E12351. I didn't have any bright ideas
for ensuring uniqueness, until we I18Nize everything and they're message
lookups. Without machine-directed output they're less interesting. If you
feel they should be more readable I can do that.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/2049/files#r38278279>.

sosiouxme replied:
Well you know, if we continued passing in message templates or formats... they could be a hash of that... :)

E012345 it is.

liggitt commented on pkg/diagnostics/README.md:
```diff
@@ -0,0 +1,113 @@
+
+Diagnostic messages are designed to be logged either for human consumption
+("text" format) or for scripting/automation ("yaml" or "json" formats). So
+messages should:
```
update with our discussion about human consumption bits

liggitt commented on pkg/diagnostics/client/config_contexts.go:
```diff
@@ -0,0 +1,289 @@
+which is not defined; it is likely that a mistake was introduced while
+manually editing your config. If this is a simple typo, you may be
+able to fix it manually.
+The master creates a fresh client config when it is started; it may be
```
the master doesn't create client configurations when started from a config file...

liggitt commented on pkg/diagnostics/client/config_contexts.go:
```diff
@@ -0,0 +1,289 @@
+needs to be updated to include the name you are using to reach it.
+
+If the master API server is generating its own certificates (which is
+default), then specify the public master address in the master-config.yaml
```
specifying

liggitt commented on pkg/diagnostics/client/config_contexts.go:
```diff
@@ -0,0 +1,289 @@
+* You may have specified the wrong host address.
+* This could mean the host is completely unavailable (down).
+* This could indicate a routing problem or a firewall that simply
+  drops requests rather than responding by reseting the connection.
```
resetting

liggitt commented on pkg/diagnostics/client/config_contexts.go:
```diff
@@ -0,0 +1,289 @@
+		errId, reason = "clientMalformedHTTP", clientMalformedHTTP
+	case strings.Contains(errMsg, "tls: oversized record received with length"):
+		errId, reason = "clientMalformedTLS", clientMalformedTLS
+	case regexp.MustCompile(`403 Forbidden: Forbidden: "/osapi/v\w+/projects?namespace=" denied by default`).MatchString(errMsg):
```
should probably match less of the message... `osapi` is even out of date now

sosiouxme replied:
Message has changed completely anyway; now it references `system:anonymous`

liggitt commented on pkg/diagnostics/client/config_contexts.go:
```diff
@@ -0,0 +1,289 @@
+	}
+	authName := context.AuthInfo
+	if _, exists := d.RawConfig.AuthInfos[authName]; !exists {
+		r.Errorf(errorKey, nil, "%s:\n Client config context '%s' has a user identity '%s' which is not defined.", unusableLine, d.ContextName, authName)
```
the name of the auth info stanza is not likely to be the user identity... it's typically a fully-qualified username-servername string

liggitt commented on pkg/diagnostics/types/error.go:
```diff
@@ -0,0 +1,40 @@
+
+// DiagnosticError is an error created by the diagnostic framework and has a little
+// more info than a regular error to make them easier to identify in the receiver.
+type DiagnosticError struct {
```
`DiagnosticMessage` since it's used for both errors and warnings?

sosiouxme replied:
It implements the error interface and is passed back as an error return value... it seems a little odd to call DiagnosticMessage{}.Error() but I'll change it if you still think it's clearer.

liggitt commented on pkg/diagnostics/systemd/unit_status.go:
```diff
@@ -0,0 +1,128 @@
+	unitRequiresUnit(r, d.SystemdUnits["openshift-node"], d.SystemdUnits["openvswitch"], sdUnitSDNreqOVS)
+	unitRequiresUnit(r, d.SystemdUnits["openshift-master"], d.SystemdUnits["openvswitch"], `Masters use openvswitch for access to cluster SDN networking`)
+	// all-in-one networking *could* be simpler, so fewer checks
+	unitRequiresUnit(r, d.SystemdUnits["openshift"], d.SystemdUnits["docker"], `Nodes use Docker to run containers.`)
```
@sdodson how frequently do unit dependencies change? any way to keep the diagnostics up to date to avoid false or missing checks?

sosiouxme replied:
So far they've changed every month or two.

Also have to update for AEP...

Eventually diagnostics could probably be part of an enterprise e2e test?

liggitt commented on pkg/cmd/experimental/diagnostics/options/flaginfo.go:
```diff
@@ -0,0 +1,55 @@
+const (
+	FlagMasterConfigName   = "master-config"
+	FlagNodeConfigName     = "node-config"
+	FlagClusterContextName = "cluster-context"
```
admin-context?

sosiouxme replied:
Projects have admins too. I thought that would be too ambiguous. Also this is used for "cluster" diagnostics. I'll change it if you think "admin" is better though.

liggitt commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -131,6 +123,20 @@ func NewCommandStartMaster(basename string, out io.Writer) (*cobra.Command, *Mas
 	return cmd, options
 }
 
+func MasterArgsAndFlags(flags *pflag.FlagSet) *MasterArgs {
```
looks like these aren't used now... can we revert the change in this file and in start_node.go?

sosiouxme replied:
yep, will save me the repeated merge conflicts on rebase too...

liggitt commented on pkg/diagnostics/cluster/registry.go:
```diff
@@ -0,0 +1,281 @@
+}
+
+func (d *ClusterRegistry) verifyRegistryImageStream(service *kapi.Service, r types.DiagnosticResult) {
+	imgStream, err := d.OsClient.ImageStreams(kapi.NamespaceDefault).Create(&osapi.ImageStream{ObjectMeta: kapi.ObjectMeta{GenerateName: "diagnostic-test"}})
```
diagnostics mutating server state is not something I would expect... that shouldn't happen by default, and should be called out very clearly

sosiouxme replied:
I did have some trepidation about this, but I can't think of any other way to test for this condition from a client, and it's really puzzling when it happens. Note that it's clearly communicated if the thing generated fails to be deleted for some reason. I'm open to suggestions...?

liggitt commented on pkg/diagnostics/cluster/registry.go:
```diff
@@ -0,0 +1,281 @@
+	scanner := bufio.NewScanner(readCloser)
+	for scanner.Scan() {
+		logLine := scanner.Text()
+		if regexp.MustCompile(`level=error msg="client error: Post http(\S+)/subjectaccessreviews`).MatchString(logLine) {
```
the URL is probably fragile... do we not want to call out all errors?

liggitt commented on pkg/diagnostics/systemd/systemd.go:
```diff
@@ -0,0 +1,287 @@
+				Interpretation: "You can safely ignore this message.",
+			},
+			{
+				Regexp: regexp.MustCompile("HTTP probe error: Get .*/healthz: dial tcp .*:10250: connection refused"),
```
I'm not sure the master does health checks on nodes any more... I think the nodes proactively check in to the master or else...

sosiouxme replied:
You're right, I'll look into whether it would help to check for something similar on the node side instead. Although node status checks already at least let you know something is wrong.

liggitt commented on pkg/diagnostics/systemd/systemd.go:
```diff
@@ -0,0 +1,287 @@
+			},
+			{
+				// user &{system:anonymous  [system:unauthenticated]} -> /api/v\\w+/services?namespace="
+				Regexp: regexp.MustCompile("system:anonymous\\W*system:unauthenticated\\W*/api/v\\w+/services\\?namespace="),
```
any URL with `?namespace=` is outdated and will never match

sosiouxme replied:
Thanks, I'll look for what these have become if they're still relevant.

liggitt commented on pkg/cmd/experimental/diagnostics/cluster.go:
```diff
@@ -0,0 +1,123 @@
+		return nil, nil, false, nil
+	} else {
+		subjectAccessReview := authorizationapi.SubjectAccessReview{
+			// we assume if you can list nodes, you're the cluster admin.
```
not a good assumption, lots of things can list nodes... other nodes, for example. options:
1. check for something only a cluster admin should have access to, like `list oauthaccesstokens`
2. check for a bogus verb on a bogus noun, that will only match the `* on *` rule that cluster admins have, like `educate dolphins`

liggitt replied:
I'm an idiot, just do a subjectaccess review for verb=* and resource=*. Though I would have laughed if `educate dolphins` had shipped

sosiouxme replied:
I was tempted :)
Thanks, this is the reason why I state assumptions that seem shaky... so a code reviewer knows to correct me.

liggitt commented:
I'm a little concerned about some of the detection logic, and how we'll ensure it stays valid. For example, regex matching of human-readable error messages, API version-specific URL patterns, etc. Without a way to validate the checks, they will match fewer and fewer real life scenarios over time and the project actually changes. I'm not sure how to test them, though... any ideas?

sosiouxme commented:
All I can think of are integration tests. Although it could be really hard to generate the real conditions under which an error is emitted, maybe in an integration test you could mock out enough stuff just to get specific errors emitted and test against the matchers that they actually get detected properly. Those tests would be super fragile but that's the point...

Other than that, getting things into QE's automated testing, and just getting dev/QE used to running it when anything funny is going on and updating when they don't work... 

Keeping up with product changes does imply an ongoing effort. And it won't be perfect. But, since this doesn't figure into the actual functioning of the system, the question to me is more how likely any individual diagnostic is to be helpful to someone at some point (versus potentially confusing or just cruft).

sosiouxme commented:
I think I've addressed comments so far, other than ripping out the complicated logging bits (which I'll commence now) and integration tests to check that pattern matches and such continue to work (which I'll at least investigate).

sosiouxme commented:
@liggitt @deads2k logging is simplified as we determined, revisions have been made according to the reviews... looking into more integration tests. Anything else blocking merging this tomorrow?

liggitt commented:
I'm ok with it. [test]

sosiouxme commented:
That's what I get for not running my own integration test first.

sosiouxme commented:
cross my fingers and [merge]

liggitt commented:
probably worth a follow-up adding a test-cmd.sh subtest (test/cmd/diagnostics.sh, etc) to invoke diagnostics against a running master

sosiouxme commented:
Will do, I just didn't know that was there.
